portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -15,17 +15,101 @@
+
+  private final static String NAME_PROPERTY = "name";
+  private final static String COMMENT_PROPERTY = "comment";
+  private final static String TIMESTAMP_PROPERTY = "timestamp";
+  private final static String UPVOTES_PROPERTY = "upvotes";
+  private final static String DOWNVOTES_PROPERTY = "downvotes";
+
+  private final static String ID_PROPERTY = "id";
+
+
+  @Override
+  protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    response.setContentType("application/json");
+    
+    
+    Entity commentEntity = new Entity(ENTITY_NAME);
+
+    String name = request.getParameter("name");
+    String commentText = request.getParameter("comment");
+    long timestamp = System.currentTimeMillis();
+    commentEntity.setProperty(NAME_PROPERTY, name);
+    commentEntity.setProperty(COMMENT_PROPERTY, commentText);
+    commentEntity.setProperty(TIMESTAMP_PROPERTY, timestamp);
+    commentEntity.setProperty(UPVOTES_PROPERTY, 0);
+    commentEntity.setProperty(DOWNVOTES_PROPERTY, 0);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    System.out.println("Added comment to datastore with " + name + " " + commentText + " on " + timestamp);
+
+    response.sendRedirect("/index.html");
+  }
+
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("Hello Hisao!");
+    response.setContentType("application/json;");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    ArrayList<Comment> receivedComments = new ArrayList<>();
+    //TODO: Change to sort by up/down vote when that is implemented.
+    Query query = new Query(ENTITY_NAME).addSort(TIMESTAMP_PROPERTY, SortDirection.DESCENDING); 
+    PreparedQuery results = datastore.prepare(query);
+
+    for(Entity entity : results.asIterable()){
+      System.out.println("Got comment from datastore with properties: " + entity.getProperties()); //Debug output
+      receivedComments.add(new Comment(entity));
+    }
+
+    Gson gson = new Gson();
+    String jsonComments = gson.toJson(receivedComments);
+    System.out.println(jsonComments);
+    response.getWriter().println(jsonComments);
   }
+
+  @Override 
+  public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException{
+    response.setContentType("application/json");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    try{
+      // Get key from request header
+      Key key = KeyFactory.createKey(ENTITY_NAME, Long.parseLong(request.getHeader(ID_PROPERTY)));
+      datastore.delete(key);
+    }catch(Exception exception){
+      System.out.println("Error when deleting comment!");
+      // Print error stacktrace - optional
+      System.out.println(exception);
+    }
```

Comment:
The effect of this `catch` block is to "swallow" the exception and hide it from the caller. In general, it is preferable to [fail fast](https://www.martinfowler.com/ieeeSoftware/failFast.pdf) and let the error get surfaced to the caller so they know that their request failed.

Reply:
I think we talked about this yesterday. So should I kill the server when this happens? I was thinking on using an alert that shows on the page when this happened. 

https://getbootstrap.com/docs/4.0/components/alerts/

Reply:
resolved on the call!